aten/src/ATen/FunctionalStorageImpl.h

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include <ATen/Tensor.h>
+
+namespace at {
+namespace functionalization {
+
+// See Note [Functionalization Pass In Core]
+
+struct ViewMeta {
+  ViewMeta(
+          std::function<Tensor(const Tensor&, int64_t)> forward,
+          std::function<Tensor(const Tensor&, const Tensor&, int64_t)> reverse,
+          int64_t out_idx = 0) :
+      forward_fn(forward),
+      reverse_fn(reverse),
+      out_index(out_idx)
+      {}
+
+  std::function<Tensor(const Tensor&, int64_t)> forward_fn;
+  std::function<Tensor(const Tensor&, const Tensor&, int64_t)> reverse_fn;
+  // See Note [out_idx in ViewMeta]
+  int64_t out_index;
+};
+
+class Alias {
+  public:
+    struct Update {
+        const at::Tensor new_val;
+        std::vector<ViewMeta> view_metas;
+    };
+    explicit Alias(const at::Tensor& base);
+    const at::Tensor& base() const;
+    size_t generation() const { return generation_; }
+    void add_update(const at::Tensor& updated_val, std::vector<ViewMeta> metas);
+    void apply_update(const Update& update);
```

Comment:
This is only used by `sync_update_operations` so make it private. Otherwise it is confusing what the difference between `add_update` and `apply_update` is.

Reply:
it actually doesn't even need to be a method; I ended up making it a helper function in the cpp file